Optimize performance of `from_xml`

HassanAkbar opened the pull request:
This PR optimizes/improves the performance of `from_xml` method as discussed here:
https://github.com/metanorma/metanorma-plateau/issues/159#issuecomment-2629878910

closes #284

HassanAkbar commented:
The CI is failing because of a failing spec of Oga Adapter where we are getting the `text` of all the children of the root instead of the content of the current root only. 
This is because we are using `Moxml` for Oga adapter in `lutaml-model.`
I have opened up an issue in `Moxml` and currently working on this: https://github.com/lutaml/moxml/issues/12
Will finalize this PR after fixing the issue in `Moxml`.

ronaldtse commented:
@HassanAkbar the Moxml PR is merged and released. Can we add some performance specs for lutaml-model so we know we don't regress? I think AI can help.

ronaldtse commented:
Seems that only rubocop is failing now?

HassanAkbar commented:
@ronaldtse, I have added specs for performance as well.

Currently, I have added specs to test for thresholds against `ips (iterations per second).`
For `Nokogiri` and `Oga`, the threshold value of ips is 5, whereas for `Ox`, it's value is 15. I have taken these values by sampling the CI runs and noting the ips, Since ips are totally runner (machine) dependent, The values will be different for each action as we have multiple OS combinations with multiple Ruby Versions so we can not decide a closer number than this and in future we would be informed by CI failure immediately if any piece of code regress the performance.

In the future, We can implement benchmarking by taking inspiration from `Opal`. It compares the performance of the code of the current branch with the code of the previous branch, and it is in a separate workflow, Sample run from Opal: https://github.com/opal/opal/actions/runs/13197609325/job/36842292771

For now, I have verified `more than 50% improvement` in the current branch (after performance improvement) by comparing the values in the CI of the main branch with the CI of the performance improvement branch.

As you can see in the latest CI run of the current PR, CI is passing according to the thresholds, but CI is failing in the current main branch (before improvement) as ips are almost half as that of after the improvement.

Failing CI run in current main (before improvement): https://github.com/lutaml/lutaml-model/actions/runs/13201339315/job/36853913684?pr=300

Passing CI run in performance improvement branch: https://github.com/lutaml/lutaml-model/actions/runs/13201195807/job/36853402947?pr=294

ronaldtse commented:
Thank you @HassanAkbar this is very nicely done. 

For adopting the Opal-like workflow, @hmdne could you help? We would also need to run lutaml-model on Opal of course since Plurimath is using lutaml-model now.

ronaldtse commented:
Hah we got the first benchmark fail with Ruby 3.1 on Windows slower than 15 ips on Ox. Let's merge this. Thank you @HassanAkbar !
